### ISSUE TITLE
Needs a 256-bit key, '504'bit given

### ISSUE THREAD

**Issue by jayesbe** [Combodo/DoctrineEncryptBundle#1]
I switched from ambta over to combodo and made the necessary code changes.. however got the error. 

My secret is 63 characters in length.

**Comment by bruno-ds** [comment]
Hello,

do you have utf8 chars in the key (or other multibyte stuff)?

I suppose this is it, because I almost not changed this part, I mean except the use of the `Combodo\DoctrineEncryptBundle\Security\SensitiveValue` class to store the value, the computation is the same : 
```php
$keyLengthOctet = mb_strlen($this->secretKey->peek(), '8bit');
        if ($keyLengthOctet !== 32) {
            throw new \Exception("Needs a 256-bit key, '".($keyLengthOctet * 8)."'bit given!");
        }
```

**Comment by jayesbe** [comment]
The ambta repo uses md5 to create a 16bit hash of the secret before it encrypts or decrypts. So actually, the original package is less secure as it uses only a 128bit key ultimately. 

However the secret should be hashed. 

I used hash('sha256', $key) to create a 32bit hash of the secret key as the password for encryption. 

However that would be a breaking change and anyone using the package already including yourself would have to migrate the encrypted values.

**Comment by jayesbe** [comment]
Also, it seems based on the events provided by doctrine, your method of adding a Decrypted property next to the encrypted property is the only way to prevent doctrine from continuously updating the field every time. 

    /**
     * @var string
     *
     * @Encrypted
     * @ORM\Column(name="api_key", type="string", length=255, nullable=true)
     */
    protected $apiKey;
    
    /**
     *
     * @var string
     */
    protected $apiKeyDecrypted;
    

This is correct and the right way based on what can be done IMO.

**Comment by bruno-ds** [comment]
Hello, since the existing code feed my needs, I do not really have time to modify it, but feel free to submit a PR I would be please to merge it if it stay compatible (or need a little work)

**Comment by jayesbe** [comment]
@bruno-ds It won't be compatible.  I have forked the Ambta repo and merged a few patches from a couple different repositories to create a migration path. It fixes all the issues outstanding including updating not executing an UPDATE if no fields have changed.   I used your code as well.. the Decrypted fields are required just as you have performed as well.

Here is my fork.  https://github.com/indinuity/doctrineencryptbundle
